src/com/ms/silverking/cloud/dht/benchmark/ycsb/GroupingSilverkingDB.java

```diff
@@ -32,173 +31,175 @@
+            System.out.flush();
+            System.exit(-1);
+          }
+        } catch (OperationException e) {
+          PutException pe;
+
+          pe = (PutException) e;
+          if (pe.getFailureCause(key) == FailureCause.INVALID_VERSION) {
+            System.out.println("Ignoring INVALID_VERSION");
+            System.err.println("Ignoring INVALID_VERSION");
             return 1;
+          } else {
+            throw new RuntimeException(pe);
+          }
         }
+      }
+      /**/
+      //System.out.println("Done waiting for: "+ key);
+      return 0;
+    } catch (PutException pe) {
+      System.out.println("Key failed: " + key);
+      return 1;
     }
-    
-    private class ReadOp {
-        final String        table;
-        final String        key;
-        final Set<String>   fields;
-        final HashMap<String, ByteIterator> values;
-        int                 returnCode;
-        private boolean     complete;
-        // FUTURE - CONSIDER MAKING A METHOD FOR THE THREADS TO WAIT FOR THE OPS TO COMPLETE
-        
-        public ReadOp(String table, String key, Set<String> fields, HashMap<String, ByteIterator> values) {
-            this.table = table;
-            this.key = key;
-            this.fields = fields;
-            this.values = values;
+  }
+
+  private class ReadOp {
+    final String table;
+    final String key;
+    final Set<String> fields;
+    final HashMap<String, ByteIterator> values;
+    int returnCode;
+    private boolean complete;
+    // FUTURE - CONSIDER MAKING A METHOD FOR THE THREADS TO WAIT FOR THE OPS TO COMPLETE
+
+    public ReadOp(String table, String key, Set<String> fields, HashMap<String, ByteIterator> values) {
+      this.table = table;
+      this.key = key;
+      this.fields = fields;
+      this.values = values;
+    }
+
+    public void waitForCompletion() {
+      synchronized (this) {
+        while (!complete) {
+          try {
+            this.wait();
+          } catch (InterruptedException ie) {
+          }
         }
-        
-        public void waitForCompletion() {
-            synchronized (this) {
-                while (!complete) {
-                    try {
-                        this.wait();
-                    } catch (InterruptedException ie) {
-                    }
-                }
+      }
+    }
+
+    public boolean waitForCompletion(TimeUnit timeUnit, int timeVal) {
+      synchronized (this) {
+        long deadlineMillis;
+
+        deadlineMillis = System.currentTimeMillis() + TimeUnit.MILLISECONDS.convert(timeVal, timeUnit);
+        while (!complete) {
+          try {
+            long curTimeMillis;
+
+            curTimeMillis = System.currentTimeMillis();
+            if (curTimeMillis < deadlineMillis) {
+              this.wait((int) (deadlineMillis - curTimeMillis));
+            } else {
+              return complete;
             }
+          } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)